dace/transformation/transformation.py

```diff
@@ -296,11 +378,65 @@ def from_json(json_obj, context=None):
         return ret
 
 
+class PatternNode(object):
+    """ 
+    Static field wrapper of a node or an SDFG state that designates it as part
+    of a subgraph pattern. These objects are used in subclasses of 
+    `Transformation` to represent the subgraph patterns.
+
+    Example use:
+    ```
+    @registry.autoregister_params(singlestate=True)
+    class MyTransformation(Transformation):
+        some_map_node = PatternNode(nodes.MapEntry)
+        array = PatternNode(nodes.AccessNode)
+    ```
+
+    The two nodes can then be used in the transformation static methods (e.g.,
+    `expressions`, `can_be_applied`) to represent the nodes, and in the instance
+    methods to point to the nodes in the parent SDFG.
+    """
+    def __init__(self, nodeclass: Type[Union[nd.Node, SDFGState]]) -> None:
+        """
+        Initializes a pattern-matching node.
+        :param nodeclass: The class of the node to match (can either be a state
+                          or a node type in a state).
+        """
+        self.node = nodeclass
+        self.match_instance: Optional[Transformation] = None
+
+    def __call__(self, sdfg: SDFG) -> Union[nd.Node, SDFGState]:
```

Comment:
can we avoid union in return values?

Reply:
Unfortunately not, for the same reasons as above

Reply:
Probably another approach can be used here: make different PatternNode implementations that accept different types. Anyway, it seems a good candidate for refactoring (maybe in another PR).